### ISSUE TITLE
Forgotten nuggets of wisdom

### ISSUE THREAD

**Issue by hsoft** [hsoft/collapseos#93]
The field of computing has been pioneered mostly by geniuses. Along its path, many awesome ideas where had, some of them were forgotten, or at least partly forgotten.

With cheap energy, ideas with a brute force (just throw more transistors at it!) approach became prevalent and took the place to what, with the perspective of the end of cheap energy, would be vastly superior approaches.

A good example of this is Forth (see #4). While we can't say it's forgotten, we can't say that its approach to computing is widespread. At least, I didn't know about it until I began digging into computing's past.

So I'm wondering: what kind of other nuggets of wisdom have been lost along the way? Being at the height of our connectivity, with internet giving us access to a lot of resources, now is the time to regroup those nuggets.

(Please, refrain from mentioning stuff like SysV and friends: yes, it's old and genius, but it's already part of the legacy of our modern computing. It's the path we took. What I'm interested in is paths we *didn't* take)

**Comment by xTibor** [comment]
Let's start at the fundamentals: ternary computers. Soviets were experimenting with this technology during the cold war. The idea is building hardware on top of base-3 logic instead of base-2 everything uses today. It was supposed to be cheaper and more energy efficient than their binary counterparts at the time. As you may have noticed, this tech saw zero adoption on the market through the decades.

(Extremely obscure stuff, hope it's still on topic.)

**Comment by piranna** [comment]
> Let's start at the fundamentals: ternary computers. Soviets were experimenting with this technology during the cold war. The idea is building hardware on top of base-3 logic instead of base-2 everything uses today. It was supposed to be cheaper and more energy efficient than their binary counterparts at the time. As you may have noticed, this tech saw zero adoption on the market through the decades.

Once i read that based on information theory, the space optimal encoding is about 2.63 states per bit, that leads to binary and ternary logic, being ternary more efficient... but being a lot more common to have elements with two than three states: light-oscurity, yes-no, up-down... That's why it got more traction.


> (Extremely obscure stuff, hope it's still on topic.)

I think it's still on topic :-)

**Comment by pgeorgi** [comment]
It's on-topic in terms of being such a nugget of wisdom, but as I understand Collapse OS, the idea is to be able to repurpose any kind of surviving hardware, and that won't usually be a ternary logic design.

On the other hand, Forth can be implemented everywhere with little effort (and its bootstrap can be made small enough that it can be entered into RAM on power-on with push-buttons with moderate effort if need be), so it seems to be directly relevant to the mission.

**Comment by Junkotron** [comment]
Reversable computing is a new hot research field since conventional computing is approaching the Landauer limit, i.e. the absolute minimum energy cost associated with loosing one bit of information. Opposing views exist on this matter.
What might make it interesting here is that constructing computer hardware reversable might enable lower power dissapation than previously possible on similar hardware.

**Comment by schierlm** [comment]
Somewhat along the lines of Forth is Scheme, which is a Lisp-like language, but trying to define most of its functions/macros in a platform independent way (in Scheme itself). Therefore, the part that actually needs to be ported is just a few slightly obscure builtin functions (like [`call/cc`](https://en.wikipedia.org/wiki/Call-with-current-continuation), which by itself may be another such nugget - and some people claim that you have only understood functional programming when you understand how `call/cc` works).

It is not as "extreme" as Forth, but (in  my opinion) more friendly to program and more readable than Forth. From memory footprint it is somewhere between Forth and Lisp (so probably already out of the scope for this project).

**Comment by gitjeff2** [comment]
@hsoft, how is this for a forgotten nugget of wisdom? [Punch and Read Intel Formatted Tape](https://archive.org/details/197712InterfaceAge/page/n151/mode/2up) extensively details an early Intel punch-tape format. Support for punched tape as a "storage format of last resort" is probably a very good idea.

In addition to 8080 assembly language source code, which is no doubt copyrighted by *Interface Age,* there is a series of logical diagrams that explain the code. These diagrams should be sufficient to form the basis of a clean implementation of a reader/writer for the format.

**Comment by piranna** [comment]
> Support for punched tape as a "storage format of last resort" is probably a very good idea.

Totally agree on that, probably it's the most persistent storage we could have. It makes sense to have a copy of CollapseOS stored on that format, and some instructions how to load it...

**Comment by f4grx** [comment]
The real question here is: What physical tape would be practical in a collapsed world? 19mm plastic scotch tape with adhesive removed? adhesive paper tape? Something else? narrow gift wrap ribbon?

Designing a tape reader/punch that would use a wide variety of tapes looks difficult. We should move this discussion to a new issue.

**Comment by piranna** [comment]
> Designing a tape reader/punch that would use a wide variety of tapes looks difficult. We should move this discussion to a new issue.

Agree, maybe also it's own repo. As I've said before, CollapseOS should not focus so much on software but a collection of resources under an umbrella project... being an easy to deploy OS one of them :-)

**Comment by masteph** [comment]
Hi folks,

just to mention it. 3D printing would help a lot in this field. You could
print your digitized version of the binary code to some sort of flat
plastic sheets. You could as well build an adapter made of plastic that
punches some sort morse encoder that programs a controller. Just some
random thoughts for this topic.

Best regards

Am Do., 23. Apr. 2020 um 12:08 Uhr schrieb Jesús Leganés-Combarro <
notifications@github.com>:

**Comment by gitjeff2** [comment]
> just to mention it. 3D printing would help a lot in this field. You could print your digitized version of the binary code to some sort of flat plastic sheets.

I think the easiest way to make punch tape would be to get a roll of masking tape or painters tape cover over the adhesive side. In a pinch, with a paper cutter, I'm sure notebook paper could be made to work as well. [DIY punch tape reader](https://www.instructables.com/id/DIY-Paper-TapePunch-Card-Maker-and-Reader/) instructions exist using easily sourced optical sensors. 

Using a 3d printer to make punch tape itself isn't really practical, I don't think. Forgive me if my math is off, but I think the data-density of punched tape is something like ~400 bytes per yard (meter) of 5-bit wide tape. That's something like ~2,600 yards (meters) of tape per MB of data. You're better off using the filament / resin to make stuff.

Also, I wanted to bring up arcane tricks to achieve UNIX-like multitasking on an 8-bit box. [GeckOS: a Unix-like 6502 operating system | VCFMW 2019](https://www.youtube.com/watch?v=jtlAOdJmeDI). It's 6502 specific, but should be helpful nonetheless. Their code is under [active development](https://github.com/fachat/GeckOS-V2) on GitHub.

Lastly, *Microcomputer Interfacing* by Bruce A. Artwick Prentice-Hall (1980) has a ton of information on design considerations for hardware that I think would be extremely relevant to this project. I don't want to link to PDF scans for copyright reasons. Everything from interface buses to [bit-sliced processors](https://en.wikipedia.org/wiki/Bit_slicing) gets some significant treatment.

**Comment by gitjeff2** [comment]
As a sort of follow-up to my notes about punched tape data storage...

The [Kansas City standard (KCS)](https://en.wikipedia.org/wiki/Kansas_City_standard) for encoding bits in pulses of audio was an early cross-platform standard. It's easy to decode even on very limited hardware. Originally, it was fixed at 300 baud. A variant called *Computer Users' Tape Standard* (CUTS) allowed for 1200 and 2400 baud respectively. This could be handy in point-to-point radio data links between systems that don't need [packet switching over the radio](https://en.wikipedia.org/wiki/Packet_radio). Also, data storage on still easily sourced cassette tapes would be useful. In a pinch, [Edison cylinders](https://en.wikipedia.org/wiki/Phonograph_cylinder) could be used as an effective means of data storage.

A standard Edison cylinder is 4-inches in length. Yes, the same length a toilet paper cylinder. Re-writable media can be made from cardboard tubes covered in wax. Reasonably "indestructible" WORM data storage can be made from plastic. So, it seems I was a bit too quick to dismiss @masteph on the notion of 3d printed storage media. @masteph assuming FDM, can something like this be reliably 3d printed? Would the lands and grooves be too small for FDM? Short of that, could "blanks" be 3d printed and heated just enough to be written to with a needle? Do you have any idea how long something like that would keep with, say, ABS or PETG instead of PLA? I realize it's not exactly [Superman Memory Crystal](https://en.wikipedia.org/wiki/5D_optical_data_storage), but years? Decades?

For context, some rough "back of the napkin" math gives...

|cylander size|runtime|data rate|data density|
|:-----------:|:-----:|--------:|-----------:|
|4"            | ~2 min |300 Bd   |~4.4 KiB    |
|4"            |~2 min |1,200 Bd |~17.5 KiB   |
|4"            |~2 min |2,400 Bd |~35.1 KiB   |

Making the CPU work to hard on I/O was a problem as far back as the early Z80-based [Chaos](http://koo.corpus.cam.ac.uk/chaos/) computer kits. While many early micros directly controlled their disk drives from the host CPU &mdash; the Apple 2 is an excellent example of this &mdash; it is possible to build fairly simple controllers that offload file I/O away from the CPU. The very "network-like" CBM IEC bus, [extensively detailed here](https://web.archive.org/web/20170116173410/https://www.zimmers.net/anonftp/pub/cbm/programming/serial-bus.pdf) (PDF), is an excellent example of that. It's worth noting that its infamously poor ~300 B/s data transfer rate was the result of a bug in the VIC 20-era that was never fixed for backwards compatibility reasons. Something inspired by the IEC bus protocol might be useful in more complex system builds with their own reasonably intelligent drive controllers.

**Comment by masteph** [comment]
Hi gitjeff2,

as far as I understand, the idea behind CollapseOS is to make sure, there
will be something left to jump start the technological development after
event X has happened. The main developer of CollapseOS makes sure there
will be some sort of bare minimum OS for chips you could find in a lot of
devices and use them to build some basic computers ... if you have plans on
how to do that somewhere in printed form while not being able to ask the
internet.

Your concern targets the problem of data storage either for initially
burning the chips with the os image or even loading applications after
starting the mostly hackish computer.

I had to read up a lot on the links you posted. After that I was wondering
how common those Edison cylinders are and what kinds of writing or reading
devices there are.
Well, at least here in Germany, I have never heard of such cylinders
before. The oldest recordings on those devices are in such a bad shape that
there is more noise than usable signal. So I guess they are not as good a
choice for the projects' purpose.

The idea of 3D printing the punch tape cards came to my mind firstly
because I own a printer and secondly because I think that the plastic might
last longer in better shape if stored correctly.
Also with respect to not knowing what kind of technology might survive
event X you should consider designing and building a machine (of course 3D
printed ;) ) that could work without the need of electricity. This is a
rather extreme scenario, but it might be that there is limited amount of
electricity that you would not want to waste on the reading mechanism but
on the computer itself, feeding it with data with a mechanical solution
that transforms the mechanically stored information into electrical
information at some sort of interface device.
For reading information from an analog medium please keep in mind the
signal to noise ratio which gets worse over time.

Typical 3D printers (FDM) have a rectangular or triangular base shape on
which you stuff single horizontal strings of plastic one over the other. My
printer has a base plate of 20cm by 20cm which is close to 8in by 8in.
So in theory you could either print a flat card with rectangular holes in
it (and a lot of them) or you could print something like a thin spiral. The
second approach is much more error prone and requires the resulting plastic
"tape" to be really thin, which would make it breakable far too easily.
Maybe you could print small sheets with a single bit line of information
that you can click onto other sheets like Lego bricks, forming something
similar like a tape.

As for the storage of the plastic I would suggest no light at all and as
low humidity as possible. In that case it should last a lot of years nearly
unchanged no matter which type of plastic you use. But I guess that would
also apply to cardboard punch cards under the same circumstances.

Please keep posting these interesting ideas as most of them really are
forgotten nuggets. It just depends on the target scenario how much they can
bring in the project in general.

Best regards,
Maik

Am Fr., 1. Mai 2020 um 15:57 Uhr schrieb gitjeff2 <notifications@github.com

**Comment by gitjeff2** [comment]
> I had to read up a lot on the links you posted. After that I was wondering how common those Edison cylinders are and what kinds of writing or reading devices there are.

They were very common 100 years back, but now Edison cylinders and the phonographs that play them are rare. The thing is, I suspect the devices are simple enough to make by hand. Though, I admit I fall short on how exactly to encode a cylinder with audio. But, I'm convinced it can be done given minimal materials and reasonable effort.

You alluded to the power issue. I had this in mind when I wrote my post. I should have mentioned it directly: early phonographs and record players were hand cranked. Volume was controlled by stuffing the flaring horn with loose cloth. This is the likely origin of the English language idiom "put a sock in it" which means to "be quiet" (imperative mood, informal register and considered slightly rude).

You're right, the error rate will be rather high due to audio artifacts and storage media degradation. Doppler-shift issues will be a problem as well. This is because simple "wind up" devices don't run at a constant speed. 

There are solutions to that higher in the stack such as something like this:

1. [Lax timing](https://en.wikipedia.org/wiki/(%CE%B5,_%CE%B4)-definition_of_limit) on the receiver's end.
2.  A very small self-dissolving archiver. A few dozen bytes of executable code, tops.
3. Repeat that 1&ndash;2 at the beginning and 1&ndash;2 at the end.
4. Build computationally simple but robust error correction into the archiver. 
5. Ensure each segment (file) can be extracted independently of the larger archive. 
6. Basically, think [tar](https://en.wikipedia.org/wiki/Tar_(computing)) (but with error correction) or [lzip](https://www.nongnu.org/lzip/lzip.html) (which is very robust) on steroids.

I expect there will be lots of storage and archival media options to suit as many different needs as possible for CollapseOS. This may or may not be one. It just seemed very much worth posting here.

> Please keep posting these interesting ideas as most of them really are forgotten nuggets.

Thanks! I do appreciate the feedback. Everyone here has been very positive and supportive about my posts. Even those who have been mildly critical &mdash; such as yourself &mdash; have still found value in the sorts of things I've been able to dig up and ideas I've had.

I have some other "forgotten nuggets" coming. But, as you see, I prefer to be certain of my source material; hence all of the links.

**Comment by gitjeff2** [comment]
## Preserving Programs on the Printed Page

Type-in programs were a thing in the 1970's and 1980's. Early ones were written in BASIC and easy enough to follow. You could learn to program simply by typing enough of these things in and playing with the code to see what worked and what didn't. There were a few major drawbacks to this approach:

First, type-in programs were tedious and data entry was error prone.

Second, as programs got more sophisticated they shifted to machine language. Practically speaking, this meant that the ML instructions were packed into incomprehensible BASIC `DATA` statements. This largely negated the propaedeutic value of the type-in programs.

To combat the high error rate of type in programs, *Compute!* had something called The [MLX](https://en.wikipedia.org/wiki/MLX_(software)). It was an efficient means of providing type-in programs to the subscriber. It worked by requiring the programmer to enter a short "automatic proofreader" program in BASIC, which only had to be done once. Then the subscriber would type-in the actual program &mdash; possibly in more than one sitting &mdash; and typoes would be flagged at the end of an erroneous line of code. Lines were always a fixed length.

The first version had a simple checksum, a rolling binary OR if I recall correctly. All input text was a series of 3 digit numbers. The problem was the checksum was too simple. Small clusters of erros in exactly the wrong combination could be missed.

The second version of the program (see *Compute! Gazette* Issue 31) had a much more sophisticated algorithm which only failed if `$00` and `$ff` were transposed. Since the algorithm is documented in the source code it could be repurposed and built upon for CollapseOS purposes.

Obviously, the type-in proofreader algorithm would either have to be ported to Forth. 

## Morse Code File Transfer

To aid in transmission over Morse Code, something like a [BIP-39 word list](https://github.com/bitcoin/bips/blob/master/bip-0039/bip-0039-wordlists.md) could be adopted. 

* We wold only need a word list 256 words long. Given that list only the first 4 letters are significant and the list has been scurbed for phonetically similar words. 
* Physical lines could be something like 8 data words and 1 parity word for checksum.
* Files could be transferred via Morse or even over voice links with a pad of paper.

Other possible encodings include: 
* [z-base-32](https://en.wikipedia.org/wiki/Base32#z-base-32) which is a standard. 
* A far more obscure but still valid option is the Sapphire-II stream cypher. A [C version](https://github.com/stamparm/cryptospecs/blob/master/symmetrical/sources/sapphire.cpp) exists. It was ported from the Pascal original.

**Comment by piranna** [comment]
Natural Language Programming. There has been several tries since the 80's like SmallTalk, The Last One and others, where the sintax of the programming language is more alike to normal english writting. For the purposses of this project it makes totally sense, because the code is easily understable by anybody. One of the most recent ones I've found is http://osmosian.com/, where they have a compiler of "Plain English Language" written in Plain English Language itself, compiling to Windows x86. It's using a EBNF grammar and it's mostly a regular programming language with a lot more optional "garbage" syntax words, so something similar could be done here. I'm not sure if it should be done as a transpiler to Forth, or as a real z80 compiler, thought...

**Comment by gitjeff2** [comment]
* [How to build your own Core Rope ROM](http://qrp.gr/coreROM/). Core Rope ROM has a data density of ~2.5 MB per cubic meter. 
* DIY [CMOS RAM](https://www.instructables.com/id/DIY-CMOS-RAM-Memory/) can be built from scratch from scavenged parts.
* Alternatively, DIY [RAM with transistors](https://www.instructables.com/id/DIY-RAM-Memory-Register-Style/) which uses more power and takes up more space. Depending upon available parts, one method may be preferable over the other.

**Comment by kr1sp1n** [comment]
Found out about magnetic amplifiers as an alternative to vacuum tubes and transistors.
Maybe this is interesting for DIY computers/devices build from scratch.

- https://en.wikipedia.org/wiki/Magnetic_amplifier
- https://www.nutsvolts.com/magazine/article/the_magnetic_amplifier

**Comment by gitjeff2** [comment]
[TinyBASIC](https://en.wikipedia.org/wiki/Tiny_BASIC) which was designed to fit into 4k of RAM. The original implementation coined the term "copyleft" and should reasonably be considered open source. There is an MIT Licensed implementation of [PaloAlto BASIC](https://github.com/pankuznetsov/Palo-Alto-BASIC-in-Ruby-less-then-in-500-lines), a modest extension of Tiny BASIC, in 500 lines of Ruby that could easily serve as a reference. In short, it should be straightforward to port this to Forth.
